File: 0070-climbing-stairs/0070-climbing-stairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        '''
        f(n = 1) = 1
        f(n = 2) 
            # take 1 step
            f(n - 1)
            # take 2 steps
            f(n - 2)
            retirm f(n - 2) + f(n - 1)

        a -> n - 2
        b -> n - 1

        for i from 3 to n, inclusively
            a, b = b, a + b

        a = 1
        b = 2
        '''
        if n == 1:
            return 1
        a, b = 1, 2
        for i in range(3, n + 1):
            a, b = b, a + b
        return b

File: 0070-climbing-stairs/0070-climbing-stairs.cpp
class Solution {
public:
    unordered_map<int, int> dp;
    int climbStairs(int n) {
        if (n == 0 || n == 1)
            return 1;
        if (dp.contains(n))
            return dp[n];
        int res = climbStairs(n - 2) + climbStairs(n - 1);
        dp[n] = res;
        return res;
    }
};

/*
f(n):
    if n == 1:
        return 1
    if in cache:
        return that
    // take 1 step
    f(n - 1)
    // take 2 steps
    f(n - 2)
    res = sum of 2 results above
    put to cache
    return res
*/